[scroll-animations] scroll timelines should clamp their current time between `0%` and `100%`.

graouts opened the pull request:
#### 2c4197c1bccc5b6bba8ceb5d5a491f6f7e923b61
<pre>
[scroll-animations] scroll timelines should clamp their current time between `0%` and `100%`.
<a href="https://bugs.webkit.org/show_bug.cgi?id=281690">https://bugs.webkit.org/show_bug.cgi?id=281690</a>
<a href="https://rdar.apple.com/138141442">rdar://138141442</a>

Reviewed by Simon Fraser.

Ensure we clip the scroll offset to be within the minimum and maximum scroll offsets for scroll timelines.
The spec doesn&apos;t call this out but this is Chrome&apos;s behavior. The issue <a href="https://github.com/w3c/csswg-drafts/issues/11033">https://github.com/w3c/csswg-drafts/issues/11033</a>
was filed to discuss this, but we should match that behavior for compatibility.

* LayoutTests/TestExpectations:
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac/TestExpectations:
* LayoutTests/webanimations/scroll-timeline-clamped-current-time-while-rubber-banding-ios-expected.txt: Added.
* LayoutTests/webanimations/scroll-timeline-clamped-current-time-while-rubber-banding-ios.html: Added.
* LayoutTests/webanimations/scroll-timeline-clamped-current-time-while-rubber-banding-mac-expected.txt: Added.
* LayoutTests/webanimations/scroll-timeline-clamped-current-time-while-rubber-banding-mac.html: Added.
* Source/WebCore/animation/ScrollTimeline.cpp:
(WebCore::ScrollTimeline::computeTimelineData const):

Canonical link: <a href="https://commits.webkit.org/285416@main">https://commits.webkit.org/285416@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/84d04d6161fc506060232dfe98321a877e34bb89

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/72577 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/48/builds/52002 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/55/builds/25375 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/76765 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/23790 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/74692 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/49/builds/59807 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/23611 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/76765 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/23790 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/75644 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/49/builds/59807 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/55/builds/25375 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/76765 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/49/builds/59807 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/55/builds/25375 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/22140 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/49/builds/59807 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/55/builds/25375 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/78436 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/87/builds/16823 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/19459 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/78436 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/86/builds/16871 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/55/builds/25375 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/78436 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-1-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/55/builds/25375 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11143 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/47800 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/2587 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/48867 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/50161 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/48612 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

nmoucht commented on Source/WebCore/animation/ScrollTimeline.cpp:
```diff
@@ -204,7 +204,14 @@ ScrollTimeline::Data ScrollTimeline::computeTimelineData(const TimelineRange& ra
+    auto rangeStart = floatValueForOffset(range.start.offset, maxScrollOffset);
+    auto rangeEnd = maxScrollOffset - floatValueForOffset(range.end.offset, maxScrollOffset);
+    if (rangeEnd > rangeStart)
+        scrollOffset = std::clamp(scrollOffset, rangeStart, rangeEnd);
```
This is not quite right. Currently, chrome clamps between 0 and maxScrollOffset, not the progress range. This is visible when there is an animation-range that shortens the progress range for the animation, which will cause the animation.currentTime to return a negative progress if the scroll position is before the active range.

graouts replied:
Right on, fixed in the latest iteration of this patch.

graouts committed:
[scroll-animations] scroll timelines should clamp their current time between `0%` and `100%`.
https://bugs.webkit.org/show_bug.cgi?id=281690
rdar://138141442

Reviewed by Simon Fraser.

Ensure we clip the scroll offset to be within the minimum and maximum scroll offsets for scroll timelines.
The spec doesn't call this out but this is Chrome's behavior. The issue https://github.com/w3c/csswg-drafts/issues/11033
was filed to discuss this, but we should match that behavior for compatibility.

* LayoutTests/TestExpectations:
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac/TestExpectations:
* LayoutTests/webanimations/scroll-timeline-clamped-current-time-while-rubber-banding-ios-expected.txt: Added.
* LayoutTests/webanimations/scroll-timeline-clamped-current-time-while-rubber-banding-ios.html: Added.
* LayoutTests/webanimations/scroll-timeline-clamped-current-time-while-rubber-banding-mac-expected.txt: Added.
* LayoutTests/webanimations/scroll-timeline-clamped-current-time-while-rubber-banding-mac.html: Added.
* Source/WebCore/animation/ScrollTimeline.cpp:
(WebCore::ScrollTimeline::computeTimelineData const):

Canonical link: https://commits.webkit.org/285416@main
